pkg/safepath/join_test.go

```diff
@@ -0,0 +1,107 @@
+package safepath
+
+import (
+	"os"
+	"path/filepath"
+	"runtime"
+	"testing"
+
+	"gotest.tools/v3/assert"
+	is "gotest.tools/v3/assert/cmp"
+)
+
+func TestJoinEscapingSymlink(t *testing.T) {
+	type testCase struct {
+		name   string
+		target string
+	}
+	var cases []testCase
+
+	if runtime.GOOS == "windows" {
+		cases = []testCase{
+			{name: "root", target: "C:\\"},
+			{name: "absolute file", target: "C:\\Windows\\System32\\cmd.exe"},
+		}
+	} else {
+		cases = []testCase{
+			{name: "root", target: "/"},
+			{name: "absolute file", target: "/etc/passwd"},
+		}
+	}
+	cases = append(cases, testCase{name: "relative", target: "../../"})
+
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			dir := t.TempDir()
+			err := os.Symlink(tc.target, filepath.Join(dir, "link"))
+			assert.NilError(t, err, "failed to create symlink to %s", tc.target)
+
+			_, cleanup, err := Join(dir, "link")
+			if err == nil {
+				cleanup()
+			}
+			assert.Assert(t, is.ErrorIs(err, ErrorEscapesBase{}))
+		})
+	}
+}
+
+func TestJoinGoodSymlink(t *testing.T) {
+	dir := t.TempDir()
+	assert.Assert(t, os.WriteFile(filepath.Join(dir, "foo"), []byte("bar"), 0o744), "failed to create file 'foo'")
+	assert.Assert(t, os.Mkdir(filepath.Join(dir, "subdir"), 0o744), "failed to create directory 'subdir'")
+
+	//assert.Assert(t, os.Symlink(filepath.Join(dir, "subdir"), filepath.Join(dir, "subdir_link_absolute")), "failed to create absolute symlink to directory 'subdir'")
```

Comment:
Turns out there's no way for `Openat2` to resolve absolute links (even if they don't escape the base directory).

We can either:
1. Don't promise that absolute (or even any) links will work as a mount source.
2. Drop the `openat2` and only use the "fallback" `Openat` implementation.

Personally, I'm in favor of 1 - no promise that absolute links work.
WDYT?

Reply:
Went with 3 - resolve the symlinks first and then traverse path without symlinks.

Reply:
I didn't really consider copying the Kubernetes implementation. 
It seems like we could be easily just use the Kubernetes [doSafeOpen](https://github.com/kubernetes/kubernetes/blob/55fb1805a1217b91b36fa8fe8f2bf3a28af2454d/pkg/volume/util/subpath/subpath_linux.go#L526-L609) as the fallback implementation. Using the whole implementation would require some major modifications (but you probably didn't mean to copy their whole `subpath` package.

I'm okay with it If everybody is fine with it and there are no legal objections (k8s is Apache 2, so probably not, but IANAL).